File: packages/shared/openclaw_mc_shared/schemas/tasks.py
"""Task-related Pydantic schemas for shared use across server and frontend codegen."""

from __future__ import annotations

from datetime import datetime
from typing import List, Optional

from pydantic import BaseModel, Field
from pydantic import UUID4

from .common import EvidenceType, TaskPriority, TaskStatus


# ---------------------------------------------------------------------------
# Evidence
# ---------------------------------------------------------------------------

class EvidenceSubmission(BaseModel):
    """A single piece of evidence submitted during a task transition."""
    type: EvidenceType
    url: str


class EvidenceRead(BaseModel):
    id: UUID4
    task_id: UUID4
    type: EvidenceType
    url: str
    submitted_at: datetime
    submitted_by: UUID4


# ---------------------------------------------------------------------------
# Task CRUD
# ---------------------------------------------------------------------------

class TaskBase(BaseModel):
    title: str
    description: Optional[str] = None
    type: str = "chore"  # bug | feature | chore
    priority: TaskPriority = TaskPriority.MEDIUM
    required_evidence_types: List[EvidenceType] = Field(default_factory=list)


class TaskCreate(TaskBase):
    project_ids: List[UUID4] = Field(default_factory=list)
    assignee_ids: List[UUID4] = Field(default_factory=list)


class TaskUpdate(BaseModel):
    title: Optional[str] = None
    description: Optional[str] = None
    type: Optional[str] = None
    priority: Optional[TaskPriority] = None
    required_evidence_types: Optional[List[EvidenceType]] = None
    project_ids: Optional[List[UUID4]] = None
    assignee_ids: Optional[List[UUID4]] = None


class TaskRead(BaseModel):
    id: UUID4
    org_id: UUID4
    title: str
    description: Optional[str] = None
    type: str
    priority: str
    status: str
    required_evidence_types: List[str] = Field(default_factory=list)
    project_ids: List[UUID4] = Field(default_factory=list)
    assignee_ids: List[UUID4] = Field(default_factory=list)
    dependency_ids: List[UUID4] = Field(default_factory=list)
    evidence: List[EvidenceRead] = Field(default_factory=list)
    completed_at: Optional[datetime] = None
    archived_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime


# ---------------------------------------------------------------------------
# Transition
# ---------------------------------------------------------------------------

class TaskTransition(BaseModel):
    """Request body for POST /tasks/{taskId}/transition."""
    to_status: TaskStatus
    evidence: List[EvidenceSubmission] = Field(default_factory=list)


# ---------------------------------------------------------------------------
# Dependencies
# ---------------------------------------------------------------------------

class DependencyAdd(BaseModel):
    """Request body for POST /tasks/{taskId}/dependencies."""
    blocked_by_id: UUID4


class DependencyRead(BaseModel):
    task_id: UUID4
    blocked_by_id: UUID4

File: frontend/src/api/tasks.ts
import api from './client'

export interface Evidence {
  id: string
  task_id: string
  type: string
  url: string
  submitted_at: string
  submitted_by: string
}

export interface Task {
  id: string
  org_id: string
  title: string
  description: string | null
  type: string
  priority: string
  status: string
  required_evidence_types: string[]
  project_ids: string[]
  assignee_ids: string[]
  dependency_ids: string[]
  evidence: Evidence[]
  completed_at: string | null
  archived_at: string | null
  created_at: string
  updated_at: string
}

export interface TaskFilters {
  project_id?: string
  status?: string
  assignee_id?: string
  priority?: string
  page?: number
  per_page?: number
}

export interface EvidenceSubmission {
  type: string
  url: string
}

export async function listTasks(orgSlug: string, filters?: TaskFilters): Promise<Task[]> {
  const { data } = await api.get(`/orgs/${orgSlug}/tasks`, { params: filters })
  return data
}

export async function createTask(orgSlug: string, body: {
  title: string
  description?: string
  type?: string
  priority?: string
  required_evidence_types?: string[]
  project_ids?: string[]
  assignee_ids?: string[]
}): Promise<Task> {
  const { data } = await api.post(`/orgs/${orgSlug}/tasks`, body)
  return data
}

export async function getTask(orgSlug: string, taskId: string): Promise<Task> {
  const { data } = await api.get(`/orgs/${orgSlug}/tasks/${taskId}`)
  return data
}

export async function updateTask(orgSlug: string, taskId: string, body: Partial<{
  title: string
  description: string
  type: string
  priority: string
  required_evidence_types: string[]
  project_ids: string[]
  assignee_ids: string[]
}>): Promise<Task> {
  const { data } = await api.patch(`/orgs/${orgSlug}/tasks/${taskId}`, body)
  return data
}

export async function transitionTask(orgSlug: string, taskId: string, body: {
  to_status: string
  evidence?: EvidenceSubmission[]
}): Promise<Task> {
  const { data } = await api.post(`/orgs/${orgSlug}/tasks/${taskId}/transition`, body)
  return data
}

export async function addDependency(orgSlug: string, taskId: string, blockedById: string): Promise<void> {
  await api.post(`/orgs/${orgSlug}/tasks/${taskId}/dependencies`, { blocked_by_id: blockedById })
}

export async function removeDependency(orgSlug: string, taskId: string, blockedById: string): Promise<void> {
  await api.delete(`/orgs/${orgSlug}/tasks/${taskId}/dependencies/${blockedById}`)
}
